libs/parallel-zk/include/nil/crypto3/zk/commitments/detail/polynomial/proof_of_work.hpp

```diff
@@ -47,37 +48,66 @@ namespace nil {
                     using transcript_type = transcript::fiat_shamir_heuristic_sequential<transcript_hash_type>;
                     using output_type = OutType;
 
-                    static inline OutType generate(transcript_type &transcript, OutType mask=0xFFFF) {
-                        output_type proof_of_work = std::rand();
-                        output_type result;
-                        std::vector<std::uint8_t> bytes(4);
+                    static inline std::array<std::uint8_t, sizeof(OutType)>
+                        int_be(OutType v) {
+                            std::array<std::uint8_t, sizeof(OutType)> bytes;
+                            for(int i = sizeof(v)-1; i>=0; --i) {
+                                bytes[i] = v & 0xFF;
+                                v >>= 8;
+                            }
+                            return bytes;
+                        }
+
+                    static inline OutType generate(transcript_type &transcript, std::size_t grinding_bits = 16) {
+                        BOOST_ASSERT_MSG(grinding_bits < 64, "Grinding parameter should be bits, not mask");
+                        output_type mask = grinding_bits > 0 ? ( 1ULL << grinding_bits ) - 1 : 0;
+                        output_type pow_seed = std::rand();
+
+                        /* Enough work for ~ two minutes on 48 cores, keccak<512> */
+                        std::size_t per_block = 1 << 30;
```

Comment:
This must somehow depend on  grinding_bits maybe? Like at least make it std::size_t per_block = 1 << min(30, grinding_bits); ???

Even though I understand it does not matter much, the cores will stop when they find a solution...